forem/forem
# Require profile fields to belong to a group

## What type of PR is this? (check all applicable)


## Description

As we move to make the profile fields option generally accessible (remove the feature flag guards), there is a behavior that no longer makes sense regarding ungrouped fields. I clarified with Jenny that these should not be possible to create (any custom field should belong to a profile field group).

Remove the empty group dropdown from the group select tag

Remove the optional: true from the belongs_to validation (raise a
validation error on save if the group is not provided)

Update test to remove belongs_to's optional setting



## Related Tickets & Documents

Internal slack discussion clarifying the intentions for this feature: 
https://forem-team.slack.com/archives/C02P0RUGS04/p1643750964238099?thread_ts=1643747780.470129&cid=C02P0RUGS04

## QA Instructions, Screenshots, Recordings


Enable the `profile_admin` feature flag at http://localhost:3000/admin/feature_flags/features 

Try to make a profile field with no group - you should receive a validation error

```ruby
 pf = ProfileField.new(label: "a label")
 pf.errors.messages
=> {:profile_field_group=>["must exist"]

 pf.save!
  TRANSACTION (1.9ms)  BEGIN
  ProfileField Exists? (1.9ms)  SELECT 1 AS one FROM "profile_fields" WHERE LOWER("profile_fields"."label") = LOWER($1) LIMIT $2  [["label", "a label"], ["LIMIT", 1]]
  TRANSACTION (1.1ms)  ROLLBACK
ActiveRecord::RecordInvalid: Validation failed: Profile field group must exist
```

Visit the admin page for profile groups http://localhost:3000/admin/customization/profile_fields

On main, there's a blank option in the group select dropdown:

![empty-group-option](https://user-images.githubusercontent.com/1237369/152056028-a17d233b-e7ab-4b9d-8d5e-40df086d7b36.png)

On this branch, that's gone:

![no-empty-dropdown-option](https://user-images.githubusercontent.com/1237369/152056296-79578d4d-5b4b-4ab3-819b-6c37036995ab.png)




### UI accessibility concerns?


## Added/updated tests?

      have not been included_

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._


## [optional] Are there any post deployment tasks we need to perform?

We may have invalid fields on forem's that have had this feature flag enabled. Those will still show (and can be edited to belong to a group or deleted, but can't be updated otherwise) in admin.

We still have the "ungrouped profile fields" partial, that should be removed once there aren't any concerns about existing invalid data. 

There's a `dependent: nulllify` condition in the ProfileFieldGroup model - should this be destroy (or is that the _other_ case where a field may have no group) or would that mean we want to keep the ungrouped profile fields as a catch-all to collect fields where the group was removed? 

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](gif_link)
